backends/ldap.go

```diff
@@ -0,0 +1,343 @@
+package backends
+
+import (
+	"fmt"
+	"github.com/go-ldap/ldap/v3"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"strconv"
+	"strings"
+)
+
+type LDAPClientFactory func(LDAP) (LDAPClient, error)
+
+type LDAPClient interface {
+	Close() error
+	Bind(username, password string) error
+	Search(request *ldap.SearchRequest) (*ldap.SearchResult, error)
+}
+
+type LDAP struct {
+	LDAPClientFactory        LDAPClientFactory
+	LDAPClient               LDAPClient
```

Comment:
Already commented on the `Factory` thing, but apart of that the struct is already named `LDAP`, its client doesn't need the prefix as well since it's private, it may simply be called called `client`.

Reply:
Sure thing